### PR TITLE
add xsync for evm and relay block

### PR DESCRIPTION
**Pull request by hench-ye**

### CODE REVIEW

File: src/xtopcom/xtxstore/xtxstoreimpl.h [hunk 1]
```diff
@@ -55,6 +55,7 @@ class xtxstoreimpl : public base::xvtxstore_t {
     mutable std::mutex m_node_type_mutex{};
     common::xnode_type_t m_combined_node_type;
     common::xbool_strategy_t m_txstore_strategy;
+    common::xbool_strategy_t m_txstore_relay_strategy;
```
Comment (jimmy1984xu) [inline src/xtopcom/xtxstore/xtxstoreimpl.h:58]: why add m_txstore_relay_strategy?

File: src/xtopcom/xblockstore/src/xvunithub.cpp [hunk 2]
```diff
@@ -731,7 +731,8 @@ namespace top
                 // return false;
             }
 
-            if (container_account->get_account() == std::string(sys_contract_relay_table_block_addr)) {
+            if (container_account->get_account() == std::string(sys_contract_relay_table_block_addr) &&
+                base::xvchain_t::instance().get_xtxstore()->check_relay_store()) {
```
Comment (jimmy1984xu) [inline src/xtopcom/xblockstore/src/xvunithub.cpp:735]: relay_block should be unpacked and stored in relay nodes now

File: src/xtopcom/xblockstore/src/xvblockhub.cpp [hunk 3]
```diff
@@ -2732,10 +2733,14 @@ namespace top
             if(this_index->check_modified_flag() == false)//nothing changed
                 return true;
 
-            const std::string key_path2 = base::xvdbkey_t::create_prunable_blockhash_key(get_cache_block_hash(this_index));
-            base::xvchain_t::instance().get_xdbstore()->set_value(key_path2, this_index->get_account() + "/" + base::xstring_utl::uint642hex(this_index->get_height()));
-            xdbg("xrelay_plugin::write_index: %s, %s, %s", this_index->get_account().c_str(), HexEncode(this_index->get_extend_data()).c_str(),
-                HexEncode(this_index->get_block_hash()).c_str());
+            if (base::xvchain_t::instance().get_xtxstore()->check_relay_store()) {
```
Comment (jimmy1984xu) [inline src/xtopcom/xblockstore/src/xvblockhub.cpp:2736]: store relayblock txs  should be provided by txstore and controlled by txstore

File: src/xtopcom/xtxstore/src/xtxstoreimpl.cpp [hunk 4]
```diff
@@ -77,6 +80,87 @@ base::xauto_ptr<base::xvtxindex_t> xtxstoreimpl::load_relay_tx_idx(const std::st
...
+        return false;
+
+    base::xvchain_t::instance().get_xdbstore()->set_value(key_path2, this_index->get_account() + "/" + base::xstring_utl::uint642hex(this_index->get_height()));
+    xdbg("xtxstoreimpl::write_relay_index, %s, %s, %s",
+         this_index->get_account().c_str(),
+         base::xstring_utl::to_hex(this_index->get_extend_data()).c_str(),
+         base::xstring_utl::to_hex(this_index->get_block_hash()).c_str());
+    return true;
+}
+bool xtxstoreimpl::store_relay_txs(base::xvblock_t * block_ptr) {
```
Comment (jimmy1984xu) [inline src/xtopcom/xtxstore/src/xtxstoreimpl.cpp:102]: should do permission control

File: src/libraries/xsync/src/xsync_object.cpp [hunk 5]
```diff
@@ -116,29 +116,56 @@ COMMANDS:\n\
...
+void xtop_sync_object::display_init(std::map<common::xenum_node_type, xsync_table_data>& table_display) const {
+    xsync_table_data table_data;
+    table_display[common::xnode_type_t::rec] = table_data;
+    table_display[common::xnode_type_t::rec].display = true;
+    table_display[common::xnode_type_t::zec] = table_data;
+    table_display[common::xnode_type_t::consensus] = table_data;
+    table_display[common::xnode_type_t::evm] = table_data;
+    table_display[common::xnode_type_t::relay] = table_data;
+}
+common::xenum_node_type xtop_sync_object::get_table_type(const std::string& table_prefix) const {
```
Comment (PaytonWu) [inline src/libraries/xsync/src/xsync_object.cpp:128]: 1. better to use common::xaccount_base_address_t.
2. common::xaccount_base_address_t makes sure it doesn't contain '@table_id' part.  
3. common::xaccount_address_t may contain '@table_id'. and common::xaccount_address_t has 'base_address()' API to get corresponding common::xaccount_base_address_t data.
4. in xnative_contract_address.h, it defines 'rec_table_address' which is equal to 'sys_contract_beacon_table_block_addr' but in type of common::xaccount_address_t
5. do not use string type anymore.

File: src/libraries/xsync/src/xsync_object.cpp [hunk 6]
```diff
@@ -116,29 +116,56 @@ COMMANDS:\n\
...
+        return common::xnode_type_t::consensus;
+    } else if (table_prefix == sys_contract_eth_table_block_addr) {
+        return common::xnode_type_t::evm;
+    } else if (table_prefix == sys_contract_relay_table_block_base_addr) {
+        return common::xnode_type_t::relay;
+    }
+    xdbg("unknown table type: %s", table_prefix.c_str());
+    return common::xnode_type_t::invalid;
+}
+std::string xtop_sync_object::get_title(common::xenum_node_type type) const {
```
Comment (PaytonWu) [inline src/libraries/xsync/src/xsync_object.cpp:143]: in xnode_type.h, it declares 'std::string to_presentation_string(xnode_type_t type);' API for getting string presentation of a concrete node type.

File: src/libraries/xsync/src/xsync_object.cpp [hunk 7]
```diff
@@ -116,29 +116,56 @@ COMMANDS:\n\
...
-        uint64_t total_shard_cur_height = 0;
-        uint64_t total_shard_max_height = 0;
-
         uint64_t total_cur_height = 0;
         uint64_t total_max_height = 0;
 
-        std::map<uint32_t, xsync_progress_t> beacon_tables_progress;
-        std::map<uint32_t, xsync_progress_t> zec_tables_progress;
-        std::map<uint32_t, xsync_progress_t> shard_tables_progress;
+        std::map<common::xenum_node_type, std::map<uint32_t, xsync_progress_t>> tables_progress;
```
Comment (PaytonWu) [inline src/libraries/xsync/src/xsync_object.cpp:165]: replace common::xenum_node_type by common::xnode_type_t

File: src/libraries/xsync/src/xsync_object.cpp [hunk 8]
```diff
@@ -248,56 +256,24 @@ std::string xtop_sync_object::status() const {
...
-                result += tmp;
-                result += "%";
-            }
-            result += "\n";
-
-            dump_chains(result, zec_tables_progress);
-        }
-
-        if (!shard_tables_progress.empty()) {
+        for (auto it:tables_progress) {
```
Comment (PaytonWu) [inline src/libraries/xsync/src/xsync_object.cpp:259]: auto const &

File: src/libraries/xsync/xrole_xips_manager.h [hunk 9]
```diff
@@ -119,6 +119,7 @@ class xrole_xips_manager_t {
     std::vector<common::xnode_address_t> get_rand_full_nodes(uint32_t max_peers);
     std::vector<common::xnode_address_t> get_full_nodes();
     std::vector<common::xnode_address_t> get_edge_archive_list();
+    std::vector<common::xnode_address_t> get_relay_list();
```
Comment (PaytonWu) [inline src/libraries/xsync/xrole_xips_manager.h:122]: const API?

File: src/xtopcom/xdata/xnative_contract_address.h [hunk 10]
```diff
@@ -51,6 +51,7 @@ XINLINE_CONSTEXPR char const * sys_contract_eth_table_block_addr_with_suffix{"Ta
 
 XINLINE_CONSTEXPR char const * sys_contract_relay_table_block_base_addr{"Ta0005"};
 XINLINE_CONSTEXPR char const * sys_contract_relay_table_block_addr{"Ta0005@0"};
+XINLINE_CONSTEXPR char const * sys_contract_relay_block_base_addr{"Tb0005"};
 XINLINE_CONSTEXPR char const * sys_contract_relay_block_addr{"Tb0005@0"};
```
Comment (PaytonWu) [inline src/xtopcom/xdata/xnative_contract_address.h:55]: why we add a 'b' type?

File: src/xtopcom/xtxstore/src/xtxstoreimpl.cpp [hunk 11]
```diff
@@ -294,5 +380,9 @@ int xtxstoreimpl::load_block_by_hash(const std::string& hash, std::vector<base::
     }
     return 0;
 }
-
+bool xtxstoreimpl::check_relay_store() {
```
Comment (PaytonWu) [inline src/xtopcom/xtxstore/src/xtxstoreimpl.cpp:383]: const API

**Review by PaytonWu — CHANGES_REQUESTED** [review]

File: src/xtopcom/xdata/xnative_contract_address.h (hunk 10 above)
Comment (PaytonWu) [reply src/xtopcom/xdata/xnative_contract_address.h:55]: i got the answer

File: src/libraries/xsync/src/xsync_object.cpp (hunk 6 above)
Comment (hench-ye) [reply src/libraries/xsync/src/xsync_object.cpp:143]: not the same

File: src/xtopcom/xdata/xnative_contract_address.h (hunk 10 above)
Comment (hench-ye) [reply src/xtopcom/xdata/xnative_contract_address.h:55]: a new type

**Pull request closed without merging**